### PR TITLE
MM-42278 Playwright base setup

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary
~~Submitted on top of https://github.com/mattermost/mattermost-webapp/pull/10020 for ease of review. See https://github.com/mattermost/mattermost-webapp/commit/23a58235f0a20e7d8f5315510a7569e289059483 for specific changes in this PR.~~ PR updated.

This PR introduced Playwright initially for visual testing, then eventually for v2 of end-to-end tests. Once it reached stability and maturity based on our use case, completed other important items like reporting, integration, pipeline, etc. and accepted for dev/community to write tests, it will be used to automate tests of new features while Cypress will go on maintenance mode.

Files and folders are to be found at `/e2e/playwright/`.

Visual testing is known to be problematic for elements that render dynamic information like date/time, name of teams/channels/users, etc. To deal with it, it will be hidden as needed using [page.addStyleTag([options])](https://playwright.dev/docs/api/class-page#page-add-style-tag) (thanks to @hmhealey for the [info](https://github.com/mattermost/mattermost-webapp/pull/9930#issuecomment-1059528475)).

I've extended `Client4` and made some changes to its properties and relocated couple of helper functions to make it compatible as node client. `Client4` was more of a browser client, though. Well noted on @hmhealey's comment [here](https://github.com/mattermost/mattermost-webapp/pull/9930#issuecomment-1059528475) and I'll update once `@mattermost/client` is released but for now, I'm proposing to change here for Playwright's setup to move forward.

For local development and in-house visual test run, screenshots were created using the official [Playwright docker container](https://playwright.dev/docs/docker) which uses Linux/Ubuntu (focal) distribution. Using such container makes the screenshot consistent even with developers using different local environment and will match with the same container we'll use in the pipeline. Screenshots generated by MacOS (darwin) or Windows are included in `.gitignore`.

To test screenshots in wide-range of different environments (OS, browser and mobile clients) and easy enough for QA to handle during release testing, I've added the ability to run both using [applitools](https://applitools.com/) and [percy](https://percy.io/). Nothing official yet but both provided so we'll get enough time to evaluate which tool is best for us.

To follow in separate PR:
- Support for remote storage of golden screenshots via S3 so that webapp repo will not be bloated with screenshot images
- Support for Cloud server
- Support with Calls product
- Reporting - to community, dashboard and Zephyr
- Gitlab pipeline
- Docs on best practices, semi-opinionated patterns and styles

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42278

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: e2e/playwright/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+## Local development
+
+#### 1. Start local server in a separate terminal.
+
+#### 2. Install dependencies and run the test.
+
+```
+# Install npm packages
+npm i
+
```
Comment (Phrynobatrachus) [inline e2e/playwright/README.md:10]: I also had to `npx playwright install`.

Comment (saturninoabril) [reply e2e/playwright/README.md:10]: Thanks Cass! Added with:
```
# Install browser binaries as prompted if Playwright is just installed or updated
# See https://playwright.dev/docs/browsers
npx playwright install
```

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Comment by hmhealey** [comment]
I tried running those login tests myself. They worked in Docker, but not locally. I'm not sure if that failure was just caused by my local server not being set up with the correct credentials though

**Review by Phrynobatrachus — APPROVED** [review]
Impressive! Both integrations work and everything looks solid to me.

**Review by devinbinnie — APPROVED** [review]
LGTM! Awesome stuff @saturninoabril 🚀

**Review by josephbaylon — APPROVED** [review]
LGTM. I was able to have a successful run using
```
npm run test -- login --project=chrome
```

I did install `npx playwright install` but I couldn't get the `npm run test -- login` to have a successful run. Wondering if it's missing an installation step. But the one with chrome above worked for me, so I'll approve the PR.

**Comment by saturninoabril** [comment]
@hmhealey @josephbaylon Thanks for testing! Would you mind sharing the error logs?

**Comment by josephbaylon** [comment]
@saturninoabril haha, I guess 3rd time's a charm. I just tried it again and it seems to be cooperating now
Screen Shot 2022-03-21 at 5 03 45 PM

File: packages/mattermost-redux/src/utils/helpers_client.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// isMinimumServerVersion will return true if currentVersion is equal to higher or than the
+// the provided minimum version. A non-equal major version will ignore minor and dot
+// versions, and a non-equal minor version will ignore dot version.
+// currentVersion is a string, e.g '4.6.0'
+// minMajorVersion, minMinorVersion, minDotVersion are integers
+export const isMinimumServerVersion = (currentVersion: string, minMajorVersion = 0, minMinorVersion = 0, minDotVersion = 0): boolean => {
```
[the remark this answers is not in the document]
Comment (saturninoabril) [reply packages/mattermost-redux/src/utils/helpers_client.ts]: I see. Done, please take a look again. Thanks!

**Comment by saturninoabril** [comment]
@hmhealey If it worked in docker, then one culprit might be due to missing snapshot/s in OS other than `linux` which is expected.
Maybe you're seeing similar error like below?
```
Error: [redacted_path]-darwin.png is missing in snapshots, writing actual.
```
If it's the case, then it should pass after the first run.
Btw, if you're experiencing a different issue, then I'll wait for the logs and see if it's something we can fix in this PR.

**Comment by saturninoabril** [comment]
Hello @amyblais, if all are good with reviews, I recommend to include this in March 30th Cloud release so it will eventually become part of v6.6. Thanks!

**Comment by saturninoabril** [comment]
Thanks! You may check `e2e/playwright/test-results` folder and see the failed screenshot. Another way is to run in headed mode by adding `--headed`, as in `npm run test -- login --project=chrome --headed`. I've set it headless by default though, at `e2e/playwright/playwright.config.ts`.

**Comment by hmhealey** [comment]
Okay, perhaps the issue was with my local server not running. I tried again with `--headed` and I got the error about the files missing. After I ran it a second time with `--headed`, it passed.

Are we expecting that we're only going to check in the linux images for now, or is that why we're running the tests in Docker? I'd be on board with just running them in Docker since I've noticed doing dev stuff inside a container does seem to simplify a lot of things

**Comment by saturninoabril** [comment]
@hmhealey Glad it's finally resolved on you.
Yes, we're expecting to check in linux images only and it's the primary reason of using Docker so that images will be consistent among developers/contributors local setup and in CI pipeline. But yes, aside from that, the container has all it needs and easier to get up and running. On the other hand, running directly on host makes it easier to see what's going on and to interact/inspect with the browser during debugging in headed mode, although those can be done as well directly via IDE or by looking at the saved screenshots/recording.
Note that linux images will typically be used in our CI pipeline but during prerelease/release testing, we'll use a third-party image diff service (percy or applitools) so that it will be easier for non-tech QA to analyze the difference between actual and expected.

**Review by hmhealey — APPROVED** [review]

**Comment by saturninoabril** [comment]
@pfltdv This is going to be part of Cloud March 30th and eventually for v6.6. Is there a workaround to get this merge? Thanks!

**Comment by pfltdv** [comment]
@saturninoabril I will investigate what we can do manually. And for `untagged-build` having successful build on CircleCI should mark that green. (Currently there is a build running) https://app.circleci.com/pipelines/github/mattermost/mattermost-webapp?branch=pull%2F10023&filter=all

**Comment by saturninoabril** [comment]
Thanks @pfltdv! I thought there's still an outstanding issue. Looks like all good now. Again, thanks!

**Comment by saturninoabril** [comment]
@amyblais Done manual cherry-pick into cloud branch.

**Pull request merged**